### DISCUSSION TITLE
How to create and contribute new augmentations?

### DISCUSSION THREAD

**Discussion by adamjstewart** [kornia/kornia#2116] [Questions] [answered]
Hi! I'm interested in creating and contributing several new augmentations. However, I'm having trouble finding any documentation on this. Can anyone point me to docs on any of the following?

1. How do I create a new augmentation? Should it subclass one of these [base classes](https://kornia.readthedocs.io/en/latest/augmentation.base.html) or `nn.Module`? I would like the resulting augmentation to be compatible with `AugmentationSequential`. I'm hoping for some docs similar to [torchvision's](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html).
2. What are the inputs and outputs of the instance methods of these base classes? They aren't documented and the example isn't very helpful.
3. How do you modify augmentations to support input, mask, etc? I need to ensure that the same augmentations are applied uniformly to both input and mask.
4. Should all augmentations have both a class and a functional form?
5. Where do I add this to Kornia? In `kornia.augmentations` or `kornia.contrib`?
6. What's the process for moving things from `kornia.contrib` to `kornia.augmentations` ?
7. Where do I add unit tests for these augmentations?

Some of the augmentations I'm interested in adding/modifying at the moment:

* RandomNCrop: like RandomCrop, but with a for loop that converts an input into N smaller patches (useful for very large images)
* PadToMultiple: like PadTo, but instead of specifying the full size, you pad to the nearest multiple of 32 (useful for semantic segmentation models and variable image sizes)
* ExtractTensorPatches: only supports input at the moment, would like to add mask.
* ExtractTensorPatches: would like to add a flag to combine N and B dimensions.

There are likely more of these to come. We at [TorchGeo](https://github.com/microsoft/torchgeo) are making heavy use of Kornia for our satellite imagery datasets and I'm trying to upstream everything we do that isn't specific to geospatial data and might be of interest to the larger computer vision community.

@ashnair1 @nilsleh @isaaccorley @calebrob6

**Comment by shijianjian** [comment, marked as answer]
> 1. How do I create a new augmentation? Should it subclass one of these [base classes](https://kornia.readthedocs.io/en/latest/augmentation.base.html) or `nn.Module`? I would like the resulting augmentation to be compatible with `AugmentationSequential`. I'm hoping for some docs similar to [torchvision's](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html).

You may refer to the `crop` related directly:
https://github.com/kornia/kornia/blob/44e43139053d70d64a699b4713adc084acd0d0a6/kornia/augmentation/_2d/geometric/crop.py#L13

Importantly, having a new random generator:
https://github.com/kornia/kornia/blob/44e43139053d70d64a699b4713adc084acd0d0a6/kornia/augmentation/_2d/geometric/crop.py#L93

Then the important methods are `apply_transform` and `inverse_transform`.

> 2. What are the inputs and outputs of the instance methods of these base classes? They aren't documented and the example isn't very helpful.

E.g. https://github.com/kornia/kornia/blob/44e43139053d70d64a699b4713adc084acd0d0a6/kornia/augmentation/_2d/geometric/crop.py#L156

input -> the input tensor
params -> generated parameters
flags -> the static parameters in `__init__`
transform -> the geometric transformation tensor if have. This is for avoiding the re-computation. It is okay if you do not have it.


> 3. How do you modify augmentations to support input, mask, etc? I need to ensure that the same augmentations are applied uniformly to both input and mask.

It is currently only handled by `AugmentationSequential`, which makes sure that the intensity augmentations are not applied on masks, and all geometric augmentations will use `nearest` sampling method. 

For non-rigid geometric augmentation support, I am planning on implementing the base class in next week or so.

> 4. Should all augmentations have both a class and a functional form?

No need.

> 5. Where do I add this to Kornia? In `kornia.augmentations` or `kornia.contrib`?
> 6. What's the process for moving things from `kornia.contrib` to `kornia.augmentations` ?

If you are planing doing it PR by PR, you may put it into `kornia.augmentattion.contrib`. Once it is finished, we move it out of contrib. If we do not issue a new release, we can do it directly. Otherwise, we need to deprecate the contrib methods for at least one release.

> 7. Where do I add unit tests for these augmentations?
`test.augmentation.test_augmentation`


> * RandomNCrop: like RandomCrop, but with a for loop that converts an input into N smaller patches (useful for very large images)
> * PadToMultiple: like PadTo, but instead of specifying the full size, you pad to the nearest multiple of 32 (useful for semantic segmentation models and variable image sizes)
> * ExtractTensorPatches: only supports input at the moment, would like to add mask.
> * ExtractTensorPatches: would like to add a flag to combine N and B dimensions.

Brilliant. Please go ahead. Let us know if any help needed.

@johnnv1 @miquelmarti

**Reply by adamjstewart** [reply]
Thanks, this is really helpful! A few follow up questions:

> ```python
> class RandomCrop(GeometricAugmentationBase2D):
> ```

How does `GeometricAugmentationBase2D` differ from `AugmentationBase2D`? I don't see it documented anywhere...

> ```python
> self._param_generator = rg.CropGenerator(size)
> ```

What is this and how do I use it? Again, I don't see it documented anywhere, and it isn't called in this class. What does it do? Does it generate the `params` argument? Is this the slice to use for cropping? If I want to crop multiple times, do I create multiple of these, or run it multiple times? How do I run it?

> Then the important methods are `apply_transform` and `inverse_transform`.

According to https://kornia.readthedocs.io/en/latest/augmentation.base.html, `generate_parameters` is also required, but I don't see this defined anywhere. I'm guessing all of these other random functions are helper functions?

> input -> the input tensor

Is this "input" or is this "input", "mask", "bbox", etc? If the former, how do I apply the same augmentation to both "input" and "mask"? If the latter, how do I know which I'm getting? It seems like augmentations should differ quite a bit depending on whether you're processing an image or a bounding box, so why is there a single function for all inputs?

> params -> generated parameters

What is a parameter, and who is generating it? What is it used for?

**Reply by adamjstewart** [reply]
> The user only needs to override: _generate_parameters_, _apply_transform_ and optionally, _compute_transformation_.

I get a `NotImplementedError` if I don't define `compute_transformation`, is this expected? What is this function supposed to do?

EDIT: looks like `compute_transformation` is required iff `params` is None.

**Reply by adamjstewart** [reply]
Trying to build a mental map of the call stack for the following code:
```python
import torch
from kornia.augmentation import AugmentationSequential, RandomCrop

input = torch.rand(1, 3, 10, 10)
mask = torch.rand(1, 10, 10)

crop = RandomCrop((5, 5))
transform = AugmentationSequential(crop, data_keys=["input", "mask"])
transform(input, mask)
```
First, the third to last line runs:

* [`RandomCrop.__init__`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/_2d/geometric/crop.py#L74)
  * [`_AugmentationBase.__init__`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/base.py#L205)
    * [`_BasicAugmentationBase.__init__`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/base.py#L47)
      * [`Module.__init__`](https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L309)

Then, the second to last line runs:

* [`AugmentationSequential.__init__`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/augment.py#L141)
  * [`ImageSequential.__init__`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/image.py#L87)
    * [`SequentialBase.__init__`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/base.py#L33)
      * [`Sequential.__init__`](https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/container.py#L84)
        * [`Module.__init__`](https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L309)

Then, the last line runs:

* [`Module.__call__`](https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L1238)
  * [`Module._call_impl`](https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L1188)
    * [`AugmentationSequential.forward`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/augment.py#L328)
      * [for loop](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/augment.py#L364) over all args (input, mask)
        * [`ImageSequential.forward`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/image.py#L363)
          * [for loop](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/image.py#L377) over all augmentations
            * [`ImageSequential.apply_to_input`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/image.py#L198)
              * [`InputApplyInverse.apply_trans`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/container/utils.py#L162)

Then, for `RandomCrop`, we run:

* [`Module.__call__`](https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L1238)
  * [`Module._call_impl`](https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L1188)
    * [`RandomCrop.forward`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/_2d/geometric/crop.py#L243)
      * [`_BasicAugmentationBase.forward`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/base.py#L160)
        * [`_AugmentationBase.apply_func`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/base.py#L242)
          * [`_AugmentationBase.compute_transformation`](https://github.com/kornia/kornia/blob/v0.6.9/kornia/augmentation/base.py#L234)

This last one isn't defined for my custom augmentation, causing the error I was seeing. I'm still not entirely sure what this function is for since there's no docstring, but at least I know how we got here.

**Reply by shijianjian** [reply]
Sorry about the old documentation. I will redo it later next year. XD

`compute_transformation` is designed for rigid affine transformations, which outputs a transformation matrix for rotation, shear, etc.

Currently, you may put with an identity matrix as what happened in `IntensityAugmentation` classes. I am working on designing the non-rigid transformation base class. We can move to that one later.

**Reply by adamjstewart** [reply]
Awesome! That gets me to my next error:
```
NotImplementedError: data_key DataKey.MASK is not implemented for ExtractTensorPatches(ExtractTensorPatches(p=1, p_batch=1.0, same_on_batch=False, window_size=(5, 5))).
```
This is the same error I was getting with the builtin `ExtractTensorPatches`, still unsure how to inform Kornia that an augmentation supports both input and mask.

**Reply by shijianjian** [reply]
If your class is inherited from `GeometricAugmentationBase2D`. Then it should work.

https://github.com/kornia/kornia/blob/3bf798ee6d5e1cfdb1dcca2fa7959e6acd3007b1/kornia/augmentation/container/utils.py#L280

**Reply by adamjstewart** [reply]
Subclassing from `IntensityAugmentationBase2D` instead of `GeometricAugmentationBase2D` seems to solve most of my issues. At the moment, I'm not trying to support augmentation inversion, so I think I can get it working with just this. But in the long run I think all of the transforms I'm trying to add are geometric in nature, so it would be nice to know how to do things properly.

Here is my (hopefully) working implementation of `ExtractTensorPatches`:
```python
class _ExtractTensorPatches(IntensityAugmentationBase2D):  # type: ignore[misc]
    """Chop up a tensor into a grid."""

    def __init__(self, window_size: Union[int, Tuple[int, int]]) -> None:
        """Initialize a new _ExtractTensorPatches instance.

        Args:
            window_size: the size of each patch
        """
        super().__init__(p=1)
        self.flags = {"window_size": window_size}

    def apply_transform(
        self,
        input: Tensor,
        params: Dict[str, Tensor],
        flags: Dict[str, Any],
        transform: Optional[Tensor] = None,
    ) -> Tensor:
        """Apply the transform.

        Args:
            input: the input tensor
            params: generated parameters
            flags: static parameters
            transform: the geometric transformation tensor

        Returns:
            the augmented input
        """
        size = flags["window_size"]
        h, w = input.shape[-2:]
        padding = compute_padding((h, w), size)
        input = extract_tensor_patches(input, size, size, padding)
        input = torch.flatten(input, 0, 1)  # [B, N, C?, H, W] -> [B*N, C?, H, W]
        return input
```
Let me know if you see anything obviously wrong other than the base class and lack of features.

Next up is `RandomNCrop`. I'm still unsure of the answer to all of these questions: https://github.com/kornia/kornia/discussions/2116#discussioncomment-4509227

**Reply by shijianjian** [reply]
Strange. The mask shall not be modified if using `IntensityAugmentationBase2D`. Can you fill up the PR then I can pull and test locally?

**Reply by adamjstewart** [reply]
Ah, that's good to know, I'll subclass `GeometricAugmentationBase2D` and copy the `compute_transformation` from `IntensityAugmentationBase2D`. The reason I was seeing that error was because I was subclassing from `AugmentationBase2D` instead of `GeometricAugmentationBase2D`.

**Reply by adamjstewart** [reply]
Why does my mask get converted from [B, H, W] to [B, 1, H, W]?

**Reply by adamjstewart** [reply]
Okay, I now have a basic implementation of RandomNCrop:
```python
from typing import Any, Dict, Optional, Tuple
import torch
from kornia.augmentation import GeometricAugmentationBase2D                              
from kornia.augmentation.random_generator import CropGenerator                           
from kornia.geometry import crop_by_indices                                              
from torch import Tensor

class RandomNCrop(GeometricAugmentationBase2D):                                         
    """Take N random crops of a tensor."""                                               
                                                                                         
    def __init__(self, size: Tuple[int, int], num: int) -> None:                         
        """Initialize a new _RandomNCrop instance.                                       
                                                                                         
        Args:                                                                            
            size: desired output size (out_h, out_w) of the crop                         
            num: number of crops to take                                                 
        """                                                                              
        super().__init__(p=1)                                                            
        self._param_generator: CropGenerator = CropGenerator(size)                       
        self.flags = {"size": size, "num": num}                                          
                                                                                         
    def compute_transformation(                                                          
        self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any]            
    ) -> Tensor:                                                                         
        """Compute the transformation.                                                   
                                                                                         
        Args:                                                                            
            input: the input tensor                                                      
            params: generated parameters                                                 
            flags: static parameters                                                     
                                                                                         
        Returns:                                                                         
            the transformation                                                           
        """                                                                              
        out: Tensor = self.identity_matrix(input)                                        
        return out                                                                       
                                                                                         
    def apply_transform(                                                                 
        self,                                                                            
        input: Tensor,                                                                   
        params: Dict[str, Tensor],                                                       
        flags: Dict[str, Any],                                                           
        transform: Optional[Tensor] = None,                                              
    ) -> Tensor:                                                                         
        """Apply the transform.                                                          
                                                                                         
        Args:                                                                            
            input: the input tensor                                                      
            params: generated parameters                                                 
            flags: static parameters                                                     
            transform: the geometric transformation tensor                               
                                                                                         
        Returns:                                                                         
            the augmented input                                                          
        """                                                                              
        out = []                                                                         
        for _ in range(flags["num"]):                                                    
            params = self._param_generator(input.size())                                 
            out.append(crop_by_indices(input, params["src"], flags["size"]))             
        return torch.cat(out)      
```
The problem is that the same transformation is not applied to both the input and mask. What's the correct way to do this? I just called `_param_generator` in `apply_transform` but that doesn't seem to reset to the same seed. Should I precompute this list and cache it somewhere?

**Reply by adamjstewart** [reply]
Another idea is to create a new custom NCropGenerator that loops over CropGenerator and returns [N, B, 4, 2] bboxes. I'm guessing `_param_generator` only gets called once and `params` are reused for all calls to `RandomNCrop` for all inputs?

**Reply by shijianjian** [reply]
You do not need to call `params = self._param_generator(input.size())` in your `apply_transform`. It will be computed and passed in by the base class.

**Reply by adamjstewart** [reply]
Yes but I need to call it N times, the default passed in to the function is only called once.

**Reply by adamjstewart** [reply]
Finally got things working by creating a new generator!
```python
import torch
from kornia.augmentation.random_generator import CropGenerator
from torch import Tensor
from typing import Dict, Tuple, Union

class NCropGenerator(CropGenerator):
    """Generate N random crops."""

    def __init__(self, size: Union[Tuple[int, int], Tensor], num: int) -> None:
        """Initialize a new NCropGenerator instance.

        Args:
            size: desired output size (out_h, out_w) of the crop
            num: number of crops to generate
        """
        super().__init__(size)
        self.num = num

    def forward(
        self, batch_shape: torch.Size, same_on_batch: bool = False
    ) -> Dict[str, Tensor]:
        """Generate the crops.

        Args:
            batch_shape: input size (b, c?, in_h, in_w)
            same_on_batch: apply the same transformation across the batch

        Returns:
            the randomly generated parameters
        """
        out = []
        for _ in range(self.num):
            out.append(super().forward(batch_shape, same_on_batch))
        return {
            "src": torch.stack([x["src"] for x in out]),
            "dst": torch.stack([x["dst"] for x in out]),
            "input_size": out[0]["input_size"],
            "output_size": out[0]["output_size"],
        }
```
I'll try to upstream all of this stuff once we finish the latest TorchGeo release.

**Reply by adamjstewart** [reply]
Note that the above doesn't work unless mask has a channel dimension. Regardless, the output seems to be converted to add a channel dimension. I have no idea why this is the case, or why my mask must be float.

**Reply by shijianjian** [reply]
We wanted it to be differentiable, so that we need floating point computations here. I will probably keep it like it is for now. Since some operations may not support integer computations well.

I will try handle the channel shape stuff in the new refactoring #2117.

**Comment by edgarriba** [comment]
As a side comment to the original request maybe it’s worth to write a small tutorial about how to contribute with new augmentations
